internal/liqonet/route-operator/symmetricRoutingOperator.go

```diff
@@ -0,0 +1,165 @@
+		return ctrl.Result{}, err
+	}
+	if added {
+		klog.Errorf("successfully added route for pod {%s} with IP address {%s} to the vxlan overlay network: %v",
+			req.String(), p.Status.PodIP)
+	}
+	return ctrl.Result{}, nil
+}
+
+// NewSymmetricRoutingOperator returns a new controller ready to be setup and started with the controller manager.
+func NewSymmetricRoutingOperator(nodeName string, routingTableID int, vxlanDevice overlay.VxlanDevice,
+	nodesLock *sync.RWMutex, vxlanNodes map[string]string, cl client.Client) (*SymmetricRoutingController, error) {
+	// Check the validity of input parameters.
+	if routingTableID > unix.RT_TABLE_MAX {
+		return nil, &liqoerrors.WrongParameter{Parameter: "routingTableID", Reason: liqoerrors.MinorOrEqual + strconv.Itoa(unix.RT_TABLE_MAX)}
+	}
+	if routingTableID < 0 {
+		return nil, &liqoerrors.WrongParameter{Parameter: "routingTableID", Reason: liqoerrors.GreaterOrEqual + strconv.Itoa(0)}
+	}
+	if vxlanDevice.Link == nil {
+		return nil, &liqoerrors.WrongParameter{Parameter: "vxlanDevice.Link", Reason: liqoerrors.NotNil}
+	}
+	return &SymmetricRoutingController{
+		Client:         cl,
+		vxlanDev:       vxlanDevice,
+		nodeName:       nodeName,
+		routingTableID: routingTableID,
+		nodesLock:      nodesLock,
+		vxlanNodes:     vxlanNodes,
+		routes:         map[string]string{},
+	}, nil
+}
+
+// addPeer adds route for the given pod. It return true when the route dose not exists
+// and is added, false if the route does already exist. An error is return if something
+// goes wrong.
+func (src *SymmetricRoutingController) addRoute(req ctrl.Request, p *corev1.Pod) (bool, error) {
+	src.nodesLock.RLock()
+	defer src.nodesLock.RUnlock()
+	nodeIP, ok := src.vxlanNodes[p.Spec.NodeName]
+	if !ok {
+		return false, fmt.Errorf("ip for node {%s} has not been set yet", p.Spec.NodeName)
+	}
+	gwIP := overlay.GetOverlayIP(nodeIP)
+	dstNet := strings.Join([]string{p.Status.PodIP, "32"}, "/")
+	added, err := routing.AddRoute(dstNet, gwIP, src.vxlanDev.Link.Attrs().Index, src.routingTableID)
+	if err != nil {
+		return added, err
+	}
+	src.routes[req.String()] = dstNet
+	return added, err
+}
+
+// delRoute removes route for the given pod. It returns true when the route exists
+// and is removed. False if the route does not exist. An error if something goes
+// wrong.
+func (src *SymmetricRoutingController) delRoute(req ctrl.Request) (bool, error) {
+	dstNet, ok := src.routes[req.String()]
+	if !ok {
+		return false, nil
+	}
+	deleted, err := routing.DelRoute(dstNet, "", src.vxlanDev.Link.Attrs().Index, src.routingTableID)
+	if err != nil {
+		return deleted, err
+	}
+	delete(src.routes, req.String())
+	return deleted, nil
+}
+
+// podFilter used to filter out all the pods that are running on the same node
+// as the operator or pods that are not ready. We do not need to configure routes
+// for pods running on same node.
+func (src *SymmetricRoutingController) podFilter(obj client.Object) bool {
```

Comment:
Does the operator correspond to the liqo-gateway?

Reply:
The current operator runs inside the `liqo-route` pod.